### PR TITLE
Refactor error handling in mutations

### PR DESCRIPTION
**Pull request by maarcingebala**
This PR is an attempt to simplify the mutations code by introducing the following changes:
- added `perform_mutation` function which is fired by `mutate` - this allows to catch validation errors that occur at any time during `perform_mutation` execution:
- removed `if errors` checks 
- removed accumulating errors in `errors` lists which were passed between various functions during a mutation execution

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

File: saleor/graphql/account/mutations.py [hunk 1]
```diff
@@ -649,7 +625,7 @@ def mutate(cls, root, info, id, type):
             raise ValueError('Unknown value of AddressTypeEnum: %s' % type)
```
Comment (akjanik) [inline saleor/graphql/account/mutations.py]: It's not changed in this PR, but isn't it validated by GraphQL itself? When you try to type query/mutation with enum values not defined in schema (in this case from here https://github.com/mirumee/saleor/blob/23208ede820d720645fe6bf2e0a5e4cda0b6f794/saleor/checkout/__init__.py) graphql should disallow process such request.

File: saleor/graphql/core/mutations.py [hunk 2]
```diff
@@ -383,38 +388,60 @@ def user_is_allowed(cls, user, ids):
         """
         return True
 
+    @classmethod
+    def mutate(cls, root, info, **data):
+        if not cls.user_is_allowed(info.context.user, data):
+            raise PermissionDenied()
+
+        count, errors = cls.perform_mutation(root, info, **data)
```
Comment (akjanik) [inline saleor/graphql/core/mutations.py]: I (maybe overreacting) see some danger here. In base case this method return one argument (`None`) and it's not overwritten in this class to return more, so if someone forgets implementing this, may get confused by `TypeError: cannot unpack non-iterable NoneType object`

File: saleor/graphql/core/mutations.py [hunk 3]
```diff
@@ -383,38 +388,60 @@ def user_is_allowed(cls, user, ids):
...
+            # catch individual validation errors to raise them later as
+            # a single error
+            try:
+                cls.clean_instance(info, instance)
+            except ValidationError as e:
+                msg = '. '.join(e.messages)
+                instance_errors.append(msg)
 
             if not instance_errors:
                 instance.delete()
```
Comment (akjanik) [inline saleor/graphql/core/mutations.py]: Is this desired by design? If I have instances list, say `[1,2,3,4,5]`, where `3` raises error, this code still deletes `1` and `2`].
Then on `3`,  it adds one entry to empty list `instance_errors` (which is instantiated to empty for every iteration ¯\_(ツ)_/¯), which is used to create message of initialized `ValidationError` (not sure if it's hold in some dynamic scope to preserve data, need to dig about it) and then `4` and `5` still gets deleted.

Comment (akjanik) [reply saleor/graphql/core/mutations.py]: I think I see how it work, not sure if we should allow to delete instances if not all are validated.

File: saleor/graphql/menu/mutations.py [hunk 4]
```diff
@@ -213,21 +209,24 @@ class Meta:
...
+            menu = cls.get_node_or_error(info, menu, field='menu')
+
+        if navigation_type == NavigationType.MAIN:
+            site_settings.top_menu = menu
+            site_settings.save(update_fields=['top_menu'])
+        elif navigation_type == NavigationType.SECONDARY:
+            site_settings.bottom_menu = menu
+            site_settings.save(update_fields=['bottom_menu'])
+        else:
+            raise AssertionError(
```
Comment (akjanik) [inline saleor/graphql/menu/mutations.py]: Why `AssertionError`? And again I see we use enums, so technically this code should never be executed.

File: saleor/graphql/shop/mutations.py [hunk 5]
```diff
@@ -92,16 +90,18 @@ class Meta:
...
 
     @classmethod
-    @permission_required('site.manage_settings')
-    def mutate(cls, root, info):
-        errors = []
+    def user_is_allowed(cls, user, input):
+        return user.has_perm('site.manage_settings')
+
+    @classmethod
+    def perform_mutation(cls, root, info):
```
Comment (korycins) [inline saleor/graphql/shop/mutations.py]: It is only visual change but personaly I prefer this way:
```
    @classmethod
    def perform_mutation(cls, root, info):
        if not settings.VATLAYER_ACCESS_KEY:
            raise ValidationError(
                'Could not fetch tax rates. Make sure you have supplied a '
                'valid API Access Key.')
        call_command('get_vat_rates')
        return ShopFetchTaxRates(shop=Shop())
```

File: saleor/graphql/core/mutations.py [hunk 6]
```diff
@@ -64,66 +81,52 @@ def _update_mutation_arguments_and_fields(cls, arguments, fields):
...
             if node is None:
-                message = "Couldn't resolve to a node: %s" % global_id
-                cls.add_error(errors, field, message)
+                raise ValidationError({
+                    field: "Couldn't resolve to a node: %s" % node_id})
         return node
 
     @classmethod
-    def get_nodes_or_error(cls, ids, errors, field, only_type=None):
+    def get_nodes_or_error(cls, ids, field, only_type=None):
```
[the remark this answers is not in the document]
Comment (akjanik) [reply saleor/graphql/core/mutations.py]: I would stick with the name till we find better one - I am out of ideas (`get_nodes_or_raise_validation_error` does not look pleased to type everywhere) and we already have `get_nodes` util.

File: saleor/graphql/order/mutations/orders.py [hunk 7]
```diff
@@ -140,17 +114,19 @@ class Meta:
...
-        errors = []
-        order = cls.get_node_or_error(info, id, errors, 'id', Order)
+    def user_is_allowed(cls, user, input):
+        return user.has_perm('order.manage_orders')
+
+    @classmethod
+    def perform_mutation(cls, root, info, id, input):
+        order = cls.get_node_or_error(info, id, only_type=Order)
 
         if not input['shipping_method']:
```
[the remark this answers is not in the document]
Comment (akjanik) [reply saleor/graphql/order/mutations/orders.py]: I think that can be the case, indeed.

File: saleor/graphql/core/mutations.py [hunk 8]
```diff
@@ -431,18 +411,48 @@ def perform_mutation(cls, root, info, ids):
...
+    @classmethod
+    def mutate(cls, root, info, **data):
+        if not cls.user_is_allowed(info.context.user, data):
+            raise PermissionDenied()
+
+        count, errors = cls.perform_mutation(root, info, **data)
+        if errors:
+            errors = validation_error_to_error_type(errors)
+            return cls(errors=errors, count=count)
+        else:
```
Comment (korycins) [inline saleor/graphql/core/mutations.py]: don't need `else`

Comment (korycins) [reply saleor/graphql/core/mutations.py]: Will it work if you leave only one `return` statement? 
```
        if errors:
             errors = validation_error_to_error_type(errors)
        return cls(errors=errors, count=count)
```
What do you think?

File: saleor/graphql/checkout/mutations.py [hunk 9]
```diff
@@ -372,34 +325,33 @@ class CheckoutShippingAddressUpdate(BaseMutation, I18nMixin):
     class Arguments:
         checkout_id = graphene.ID(description='ID of the Checkout.')
         shipping_address = AddressInput(
+            required=True,
```
Comment (korycins) [inline saleor/graphql/checkout/mutations.py]: is it expected?

File: saleor/graphql/core/mutations.py [hunk 10]
```diff
@@ -154,6 +155,35 @@ def construct_instance(cls, instance, cleaned_data):
...
+        return True
+
+    @classmethod
+    def mutate(cls, root, info, **data):
+        if not cls.user_is_allowed(info.context.user, data):
+            raise PermissionDenied()
+
+        try:
+            response = cls.perform_mutation(root, info, **data)
+            if response.errors is None:
```
Comment (korycins) [inline saleor/graphql/core/mutations.py]: So... maybe `perform_mutation` should return empty errors as a `[]`?

**Pull request merged**